## Lab 03.01 - Introduction to Pandas and Data Operations
In this lab, you'll learn how to use the pandas library to analyze real-world sales data. By the end of this lab, you'll be comfortable with the fundamental operations in pandas.

### Part 0 - Intro to Pandas
Before we dive into using pandas, let's understand what it is and why it's important in data science.

To answer the questions, edit the markdown cell and put your answer below the question. 

**Make sure to save the markdown cell, by pressing the ✓ (check) icon in the top right after answering the questions**

Research and critically think about the following questions:

##### Question 00
What is Pandas, and what are its primary functions in data visualization for data science?
- **Answer:** Pandas is software library that is written in python for data manipulation and analysis.

##### Question 01
What does "data manipulation" mean in the context of pandas and data science field?
- **Answer:**Data manipulation is when you are organizing or arranging data in order to make it easier to interpret.

##### Question 02
What kind of files can pandas read? List at least 3:
- **Answer:**csv, excel, and sql.

#### 0.0 - Hands-On Exploration
Let's start by importing pandas and exploring its basic functionality. Type these commands and write down what you observe:

First, let's install the matplotlib library, in your terminal run the following command:

```bash
pip3 install pandas
```

Once installed, let's start by importing pandas and exploring its basic functionality. 

In [2]:
import pandas as pd

Now lets run create a list, and see what happens when we use it with pandas.

In [3]:
# Create a simple list of numbers
numbers = [1, 2, 3, 4, 5]

# Convert it to a pandas Series
series = pd.Series(numbers)

# Print the result
print(series)

0    1
1    2
2    3
3    4
4    5
dtype: int64


##### Question 03
What did you notice about the output format?
- **Answer:**there is the index on the left and the number on the right.

##### Question 04
How is this different from a regular Python list? What are the numbers on the left side?
- **Answer:**this is different from a regular python list because it doesn't tell you the index number

### Part 1 - Reading and Exploring Data

#### 01.00 - Loading Data
Let's start by loading our sales data that you downloaded alongside this lab. In pandas, we can read various file formats, but CSV (Comma-Separated Values) is one of the most common.

In [8]:
sales_df = pd.read_csv('sales_data.csv')

When using pandas to load data, in ingests it into what is known as a "dataframe", this allows us to do advanced manipulation.

#### 01.01 - Viewing Data Samples
Let's explore the fundamental methods for understanding our data structure and content. We'll look at each method individually and understand what it tells us about our data.

In [6]:
print(sales_df.head())

         Date    Product         Category    Price  Units    Total  Discount  \
0  2023-01-01   Notebook  Office Supplies    18.33      2    35.56      0.03   
1  2023-01-01     Laptop      Electronics  1943.35      4  7773.40      0.00   
2  2023-01-01     Tablet      Electronics   436.05      2   872.10      0.00   
3  2023-01-01  USB Drive      Accessories   116.48      3   349.44      0.00   
4  2023-01-01    Pen Set  Office Supplies    42.14      4   168.56      0.00   

  Region      Sales_Rep Payment_Method Customer_Segment  
0  South   Sarah Wilson     Debit Card   Small Business  
1  North     John Smith     Debit Card   Small Business  
2   East    David Brown  Bank Transfer       Government  
3   East  Lisa Anderson         PayPal       Individual  
4   East    David Brown     Debit Card        Corporate  


In [7]:
print(sales_df.tail())

            Date       Product         Category    Price  Units    Total  \
3766  2023-12-31    Smartwatch      Electronics  1730.72      2  3461.44   
3767  2023-12-31          Desk        Furniture   462.57      1   462.57   
3768  2023-12-31        Laptop      Electronics   560.79      2  1121.58   
3769  2023-12-31  Office Chair        Furniture   116.04      2   232.08   
3770  2023-12-31      Notebook  Office Supplies    18.15      4    72.60   

      Discount Region        Sales_Rep Payment_Method Customer_Segment  
3766       0.0  South  Michael Johnson     Debit Card       Individual  
3767       0.0  North       John Smith     Debit Card       Government  
3768       0.0   East      David Brown     Debit Card   Small Business  
3769       0.0  North       Emma Davis    Credit Card        Corporate  
3770       0.0  North       Emma Davis         PayPal       Individual  


##### Question 04
What is the difference between the `tail()` and `head()` commands? What would be use of either command be?
- **Answer:**the difference between the two is that head shows you the first 5 and tail shows you the last 5.

##### Question 05
What happens when you put a number in the `head()` function? What changed?
- **Answer:** its show the you the number you put in from the head or tail.

#### 01.02 - Understanding DataFrame Structure
Let's examine the basic properties of our DataFrame:

In [9]:
print(sales_df.columns)

Index(['Date', 'Product', 'Category', 'Price', 'Units', 'Total', 'Discount',
       'Region', 'Sales_Rep', 'Payment_Method', 'Customer_Segment'],
      dtype='object')


In [11]:
print(sales_df.dtypes)

Date                 object
Product              object
Category             object
Price               float64
Units                 int64
Total               float64
Discount            float64
Region               object
Sales_Rep            object
Payment_Method       object
Customer_Segment     object
dtype: object


##### Question 06
Compare and contrast the outputs of the `columns` and `dtypes` properties. What is similar what is different?
- **Answer:**columns shows you the names of the columns and dtypes shows you the names of the rows.

##### Question 07
What data type is the 'Date' column? Is this what you expected? 
- **Answer:**It's a object. I thought it would be a string.

##### Question 08
Why might pandas choose different data types for different columns?
- **Answer:**they might choose different data types since there might being a value that's a number or and boleen

#### 01.03 - Data Information Summary
The `info()` method provides a concise summary of our DataFrame:

In [12]:
print(sales_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3771 entries, 0 to 3770
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              3771 non-null   object 
 1   Product           3771 non-null   object 
 2   Category          3771 non-null   object 
 3   Price             3771 non-null   float64
 4   Units             3771 non-null   int64  
 5   Total             3771 non-null   float64
 6   Discount          3771 non-null   float64
 7   Region            3771 non-null   object 
 8   Sales_Rep         3693 non-null   object 
 9   Payment_Method    3739 non-null   object 
 10  Customer_Segment  3771 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 324.2+ KB
None


##### Question 09
What information does `info()` tell us, how could this information be useful?
- **Answer:** it tell's us all about how many columns there are. this information would be useful if you needed to know if how many columns there are.

##### Question 10
How much memory is our DataFrame using? 
- **Answer:** it is using 324.2+ KB

##### Question 11
What does "null" mean? What impact could having "null" values in a dataset?
- **Answer:**

#### 01.04 - Numerical Summaries
The `describe()` method provides statistical summaries for numerical columns:

In [ ]:
print(sales_df.describe())

##### Question 12
What statistics does the `describe()` function give us?
- **Answer:**

##### Question 13
Which columns did the `describe()` function perform statistics on? Why? (Hint: why didnt we see 'Product' in the output?)
- **Answer:**

In [ ]:
print(sales_df[['Price', 'Units']].describe())

##### Question 14
What is different in the codeblock above compared to the previous one. How did this affect the output?
- **Answer:**

In [ ]:
print(sales_df['Category'].unique())

##### Question 15
How many unique product categories are in our data set?
- **Answer:**

In [ ]:
print("Output 1:")
print(sales_df['Category'].value_counts())

print('')
print('')

print("Output 2:")
print(sales_df['Category'].value_counts(normalize=True))

##### Question 16
What is the difference between the two outputs?
- **Answer:**

##### Question 17
What is our most common product category?
- **Answer:**

##### Excercise 00
Using what you've learned, create a code cell direclty below this one to answer these questions:

- What is the date range of our sales data? (Hint: try `min()` and `max()` functions)
- How many unique products do we sell?
- What is the average price by category?
- What is our most common payment method?

### Part 2 - Data Selection and Filtering

#### 02.01 - Column Selection
There are multiple ways to select columns in pandas. Let's explore them:

In [ ]:
# Select single column
prices = sales_df['Price']

# Select multiple columns
product_info = sales_df[['Product', 'Price', 'Units']]

##### Question 17
Compare and contrast how we select multiple columns vs a single column.
- **Answer:**

##### Excercise 01
Using what you've learned, create a code cell direclty below this one to select the 'Region' and 'Category' columns and store them in a variable called 'location_data'.

#### 02.01 - Filtering Data
Let's learn how to filter our data based on conditions:

In [ ]:
expensive_items = sales_df[sales_df['Price'] > 500]

north_sales = sales_df[sales_df['Region'] == 'North']

print("Expensive items:")
print(expensive_items)
print("\nNorth region sales:")
print(north_sales)

##### Question 18
What is `sales_df['Price'] > 500` accomplishing in our code? How might we use this in data science workflows?
- **Answer:**

##### Question 19
What logic operator would we use to combine multiple conditonals when filtering columns?
- **Answer:**

##### Excercise 02
Using what you've learned, create a code cell direclty below this one to create the following filters:
- The number of units sold was greater than 5
- The product category is 'Electronics'

#### Part 3 - Final Summary
Use your experience completing this lab, and the code block below to answer the following summary questions

##### Question 20
In your own words, explain what pandas is and why it's useful for data analysis:
- **Answer:**

##### Question 21
How many unique products are in the dataset?
- **Answer:**

##### Question 22
What was the most popular product category in the `North` region? 
- **Answer:**